### PR TITLE
support "private" in ssb-ql-0

### PR DESCRIPTION
**Pull request by staltz**
- [x] Pending on https://github.com/ssb-ngi-pointer/ssb-subset-replication-spec/pull/17 merged first

### CODE REVIEW

**Comment by staltz** [comment]
@mycognosist It's worth taking a bit more critical perspective when reviewing this PR since we're touching private message stuff. The spec changes are at https://github.com/ssb-ngi-pointer/ssb-subset-replication-spec#query-language

File: ql0.js [hunk 1]
```diff
@@ -25,13 +25,26 @@ function validate(query) {
...
+      throw new Error(
+        `query "${k}" should be a string or null: ${JSON.stringify(obj)}"`
+      )
+    } else if (k === 'private' && typeof obj[k] !== 'boolean') {
+      throw new Error(
+        `query "${k}" should be a boolean: ${JSON.stringify(obj)}`
+      )
     }
   }
+  if (obj.private && obj.type) {
```
Comment (mycognosist) [inline ql0.js:43]: This seems to be the critical check to ensure we're not leaking metadata. I don't know all the eccentric JS type computations but I can't see any obvious way this check could be fooled; `private` has been validated for `boolean` type and `obj.type` will be truthy if it's not `null` (or `undefined` etc.).

File: ql0.js [hunk 2]
```diff
@@ -68,10 +81,15 @@ function parse(query) {
 function toOperator(query, dedicated = false) {
   validate(query)
   const actualQuery = parse(query)
-  return and(
-    author(actualQuery.author, { dedicated }),
-    type(actualQuery.type, { dedicated })
-  )
+  if (actualQuery.private) {
+    return and(author(actualQuery.author, { dedicated }), isPrivate())
```
Comment (mycognosist) [inline ql0.js:85]: 👍🏻

File: test/ql0.js [hunk 3]
```diff
@@ -8,63 +8,93 @@ const BOB_ID = ssbKeys.generate().id
...
     },
-    /"type" should be a valid string/,
+    /"type" should be a string/,
     'type not string'
   )
 
   t.throws(
     () => {
-      QL0.validate({ author: '@fafa', type: 'vote' })
+      QL0.validate({ author: ALICE_ID, type: 'vote', private: 'yes' })
```
Comment (mycognosist) [inline test/ql0.js:89]: Maybe you could add a similar `QL0.validate` test where `private: true` and `type: 'vote'`. That would ensure validation is correctly preventing metadata leaks when the `private` field is set to `true` and `type` is not `null`.

**Review by mycognosist — APPROVED** [review]
A minor test recommendation (I don't think the case I proposed is currently covered).

Otherwise, it looks solid to me. I also like the clarifying wording added to the spec in https://github.com/ssb-ngi-pointer/ssb-subset-replication-spec/pull/17

**Comment by staltz** [comment]
Thanks for taking a look! I added the test you suggested :)

**Pull request merged**